Regularization to Prevent Overfitting (#1461)
section-engineering-education/engineering-education: articles/regularization/index.md

### Lasso Regression

Lasso regression is a regularization technique to reduce model complexity. It is also known as L1 regularization. Lasso stands for Least Absolute Shrinkage and Selector Operator. Let’s look at the equation below.

$$ \sum\limits_{i=1}^n( y_i – \beta_0 - \sum\limits_{j=1}^p {\beta_j x_ij )^2} + \lambda \sum\limits_{j=1}^p \mid \beta_j \mid = RSS + \lambda \sum\limits_{j=1}^p \mid \beta_j \mid $$

We note that it is a variation of the loss function equation, with the introduction of a penalty term. To penalize high coefficients, lasso uses absolute values of beta $ (\mid \beta\mid) $. Whereas, as we shall learn, ridge regression uses the squares of beta $(\beta^2)$ for a similar purpose.
Lasso shrinks the regression coefficients to regularize the model parameters. Unlike ridge regression, as we shall cover later, lasso can reduce some regression coefficients to zero. This is particularly important when it comes to feature selection.

> The feature selection process is attributed to the ability of lasso to reduce some regression coefficients to zero. It comes after the regression coefficients are shrunk. Every non-zero regression coefficient is selected for use in the model. Since some features are eliminated, less important features will not feature in the final model. This greatly assists in minimizing prediction errors.

Reviewer: `The feature selection process is attributed to the ability of lasso to reduce some regression coefficients to zero`. 

You have mentioned this point previously.  Here, you can explain what `Feature selection` is?  and what is the significance when the regression coefficients are reduced to zero?

`It comes after the regression coefficients are shrunk. Every non-zero regression coefficient is selected for use in the model. Since some features are eliminated, less important features will not feature in the final model. This greatly assists in minimizing prediction errors.`

This whole paragraph looks unclear. Please rephrase this paragraph for improving the understandability.

> Lasso also improves the prediction accuracy of models. The shrinking of coefficients minimizes the bias and reduces the variance of models. A context that favors the use of lasso is when we have a high number of features and a small amount of data. The tuning parameter $\lambda$ controls the shrinkage. From the equation, when $\lambda$ is zero, the equation is reduced to the linear regression loss function equation. The greater the value of $\lambda$, the greater the reduction of the coefficients towards zero.

Reviewer: ```suggestion
Lasso also helps improving the prediction accuracy of the models. The shrinking of coefficients minimizes the bias and reduces the variance of models. A context that favors the use of lasso is, when we have a small dataset with more number of features. The tuning parameter $\lambda$ controls the shrinkage. From the equation, when $\lambda$ is zero, the equation is reduced to the linear regression loss function equation. The greater the value of $\lambda$, the greater the reduction of the coefficients towards zero.
```

`The shrinking of coefficients minimizes the bias and reduces the variance of models.`

Fyi, Bias and Variance are inversely proportional.

### Ridge Regression

> Ridge regression refers to a type of linear regression where in order to get better predictions in the long term, we introduce a small amount of bias. Known as L2 regularization, it is a technique used to reduce model complexity. In ridge regression, the loss function is altered. A penalty term is added to it, as we shall see below. Let’s explain ridge regression better using the following equation.

Reviewer: ```suggestion
Ridge regression refers to a type of linear regression where in order to get better predictions in the long term, we introduce a small amount of bias. It is also known as L2 regularization. In ridge regression, we have the same loss function with a slight alteration in the penalty term, as shown below:
```

$$ \sum\limits_{i=1}^n( y_i – \beta_0 - \sum\limits_{j=1}^p {\beta_j x_ij )^2} + \lambda \sum\limits_{j=1}^p \beta_j^2 = RSS + \lambda \sum\limits_{j=1}^p \beta_j^2 $$

> Above is the ridge regression equation. AS we can see, the difference between the above equation and the general equation for the loss function is that $ \lambda \sum\limits_{j=1}^p \beta_j^2 $ is added on both sides. A shrinkage quantity is added to RSS, thus modifying it. To estimate the coefficients, the function above has to be minimized. A penalty term is introduced. To penalize the flexibility of our model, we use a tuning parameter that decides the extent of the penalty. The tuning parameter is $\lambda$. Higher values of the coefficients represent a model with greater flexibility. To minimize the function, these coefficients should be small. L2 regularization ensures the coefficients do not rise too high.

Reviewer: ```suggestion
As we can see, the main difference between the above equation and the general equation for the loss function is that $ \lambda \sum\limits_{j=1}^p \beta_j^2 $ contains the squared value of the regression coefficients. The tuning parameter is $\lambda$. Higher values of the coefficients represent a model with greater flexibility. To penalize the flexibility of our model, we use a tuning parameter that decides the extent of the penalty. To minimize the function, these coefficients should be small. L2 regularization ensures the coefficients do not rise too high.
```
Most of your points are redundant.

` To minimize the function`.

Please be specific on what are you trying to minimize - loss or error or coefficients?

> As can be noted from the equation, if the value of $\lambda$ tends to zero, the equation becomes the loss function of the linear regression model. The penalty term will have no effect. This means that if the value of $\lambda$ is at a minimum, the model will be similar to a linear regression model. Conversely, if the $\lambda$ value tends to infinity, the effect of the shrinkage penalty increases. As a result, the coefficient estimates of the ridge regression will approach zero. This underlines the importance of choosing a good value of $\lambda$.

Reviewer: ```suggestion
When the value of $\lambda$ tends to zero, the L2 regularization equation becomes the loss function of the linear regression model. The penalty term will have no effect. This means that if the value of $\lambda$ is minimum, the model can be called a simple linear regression model. Conversely, when the $\lambda$ value tends to infinity, the effect of the shrinkage penalty increases. As a result, the coefficient estimates of the ridge regression will approach zero. This underlines the importance of choosing a good value for $\lambda$.
```

> Linear or polynomial regression will likely prove unsuccessful if there is high [collinearity](https://machinelearningmind.com/2019/10/19/multicollinearity-how-to-fix-it/) between the independent variables. Ridge regression is a potential solution to handle multicollinearity. We shall explore this in a different article.

Reviewer: It would be great, if you could add a single line explaining what collinearity is, and how it helps solving the problem of multicollinearity.

> However, ridge regression has one clear disadvantage; model interpretability. When shrinking the coefficients of the predictors of the least importance, it will get them to be very close to zero. Here’s the problem. They won’t actually be zero. As a result, one may end up including all the coefficients in the final model. Whereas lasso regression manages to force some coefficient estimates to zero when $\lambda$ is large enough. Resultantly, we can say lasso performs feature selection. Furthermore, the resulting models can be considered sparser compared to the ones resulting from the L2 technique. This is because some of the estimate coefficients can be truly zero. The number of predictors is reduced.

Reviewer: ```suggestion
However, Ridge regression has one clear disadvantage: model interpretability. When shrinking the coefficients of the predictors of the least importance, it will reduce very close to zero. Here’s the problem. As a result, one may end up including all the coefficients in the final model. Whereas lasso regression manages to force some coefficient estimates to zero, when the $\lambda$ is large enough. Alternatively, we can say lasso performs feature selection. 

Furthermore, the resulting models can be considered sparser when compared to the ones resulting from the L2 technique. This is because some of the estimate coefficients can be truly zero. The number of predictors is reduced.
```
A few suggestions to improve on:
- Could you please explain what you mean by `coefficients of the predictors of the least importance`?
- `including all the coefficients in the final model`. What would happen if all the coefficients are included in the final model?
- ` models can be considered sparser when compared`. Does `sparser` signify prevention of overfitting?
- `The number of predictors is reduced.` Is this sentence a continuation to the previous sentence?

### Wrapping Up

> Regularization is a formidable technique to prevent overfitting. There are several regularization techniques. We have looked at two of the most popular ones. Through this article, we have understood how the two compare to each other and how they are derived from the linear regression loss function equation. Until next time, good luck!

Reviewer: ```suggestion
Regularization is a formidable technique to prevent overfitting. There are several regularization techniques. We have looked at two of the most popular ones. Through this article, we have understood how the two techniques help in preventing overfitting and how they are derived from the linear regression loss function equation. Until next time, good luck!
```
